components/brave_shields/adblock/rs/src/engine.rs

```diff
@@ -0,0 +1,202 @@
+}
+
+pub fn read_list_metadata(list: &CxxVector<u8>) -> FilterListMetadataResult {
+    || -> Result<FilterListMetadata, InternalError> {
+        Ok(adblock::lists::read_list_metadata(std::str::from_utf8(list.as_slice())?).into())
+    }()
+    .into()
+}
+
+fn convert_cxx_string_vector_to_string_collection<C>(
+    value: &CxxVector<CxxString>,
+) -> Result<C, Utf8Error>
+where
+    C: FromIterator<String>,
+{
+    value.iter().map(|s| s.to_str().map(|t| t.to_string())).collect()
+}
+
+impl Engine {
+    pub fn enable_tag(&mut self, tag: &CxxString) -> EmptyTupleResult {
+        || -> Result<(), InternalError> { Ok(self.engine.enable_tags(&[tag.to_str()?])) }().into()
+    }
+
+    pub fn disable_tag(&mut self, tag: &CxxString) -> EmptyTupleResult {
+        || -> Result<(), InternalError> { Ok(self.engine.disable_tags(&[tag.to_str()?])) }().into()
+    }
+
+    pub fn tag_exists(&self, key: &CxxString) -> BoolResult {
+        || -> Result<bool, InternalError> { Ok(self.engine.tag_exists(key.to_str()?)) }().into()
+    }
+
+    pub fn matches(
+        &self,
+        url: &CxxString,
+        hostname: &CxxString,
+        source_hostname: &CxxString,
+        request_type: &CxxString,
+        third_party_request: bool,
+        previously_matched_rule: bool,
+        force_check_exceptions: bool,
+    ) -> BlockerResultResult {
+        || -> Result<BlockerResult, InternalError> {
+            let inner_blocker_result = self.engine.check_network_urls_with_hostnames_subset(
+                url.to_str()?,
+                hostname.to_str()?,
+                source_hostname.to_str()?,
+                request_type.to_str()?,
+                Some(third_party_request),
+                previously_matched_rule,
+                force_check_exceptions,
+            );
+            Ok(inner_blocker_result.into())
+        }()
+        .into()
+    }
+
+    pub fn get_csp_directives(
+        &self,
+        url: &CxxString,
+        hostname: &CxxString,
+        source_hostname: &CxxString,
+        request_type: &CxxString,
+        third_party_request: bool,
+    ) -> StringResult {
+        || -> Result<String, InternalError> {
+            Ok(self
+                .engine
+                .get_csp_directives(
+                    url.to_str()?,
+                    hostname.to_str()?,
+                    source_hostname.to_str()?,
+                    request_type.to_str()?,
+                    Some(third_party_request),
+                )
+                .unwrap_or_default())
+        }()
+        .into()
+    }
+
+    pub fn deserialize(&mut self, serialized: &CxxVector<u8>) -> EmptyTupleResult {
+        self.engine.deserialize(serialized.as_slice()).map_err(|e| InternalError::from(e)).into()
+    }
+
+    pub fn add_resource(
+        &mut self,
+        name: &CxxString,
+        content_type: &CxxString,
+        data: &CxxString,
+    ) -> EmptyTupleResult {
```

Comment:
(note: pretty sure we don't actually use this function anywhere)

The inputs would only ever be entered from a text `<input>` in the browser, which... appears to support pasting arbitrary binary data from the clipboard :thinking:.

I still can't imagine having code to handle specific errors here any time soon, so a "success" `bool` would be sufficient too.

Reply:
We may want to reconsider keeping the result type if multiple errors can be returned (`Utf8Error` and `AddResourceError`). If there are multiple types of potential errors, it may be useful to know why the function failed after the fact. Returning the error message could make sense in this case, since it'd be more stylish to perform the logging in cpp land with Chromium logging macros. lmk what you think

Reply:
There's still a lot that needs to be implemented before that'll be useful at all. I guess I don't really mind either way, the small amount of extra code here probably wouldn't be too much of an additional maintenance burden.

Reply:
will keep this in case it becomes useful for debugging in the future, given it's low cost